projects/plugins/boost/app/assets/src/js/pages/settings/sections/Score.svelte

```diff
@@ -5,40 +5,120 @@
-	import { getScoreLetter, clearCache, requestSpeedScores } from '../../../api/speed-scores';
+	import {
+		getScoreLetter,
+		requestSpeedScores,
+		didScoresImprove,
+		getScoreImprovementPercentage,
+	} from '../../../api/speed-scores';
+	import debounce from '../../../utils/debounce';
 	import { __ } from '@wordpress/i18n';
+	import { criticalCssStatus } from '../../../stores/critical-css-status';
+	import { modules } from '../../../stores/modules';
+	import { derived } from 'svelte/store';
+	import RatingCard from '../elements/RatingCard.svelte';
 
 	const siteIsOnline = Jetpack_Boost.site.online;
 
 	let loadError;
 	let isLoading = siteIsOnline;
+	let showPrevScores;
 	let scoreLetter = '';
 	let scores = {
-		mobile: 0,
-		desktop: 0,
+		current: {
+			mobile: 0,
+			desktop: 0,
+		},
+		previous: null,
 	};
+	let showRatingCard = false;
+	let improvementPercentage = 0;
 
 	if ( siteIsOnline ) {
 		refreshScore( false );
 	}
 
+	/**
+	 * Check if module states are currently in sync with server.
+	 */
+	const modulesInSync = derived( modules, $modules => {
+		const slugs = Object.keys( $modules );
+		for ( let i = 0; i < slugs.length; i++ ) {
+			if ( $modules[ slugs[ i ] ].synced === false ) {
+				return false;
+			}
+		}
+
+		return true;
+	} );
+	/**
+	 * String representation of the current state that may impact the score.
+	 *
+	 * @type {Readable<string>}
+	 */
+	const scoreConfigString = derived(
+		[ modules, criticalCssStatus ],
+		( [ $modules, $criticalCssStatus ] ) =>
+			JSON.stringify( {
+				modules: $modules,
+				criticalCss: {
+					created: $criticalCssStatus.created,
+				},
+			} )
+	);
+
+	/**
+	 * The configuration that led to latest speed score.
+	 *
+	 * @type {Readable<string>}
+	 */
+	let currentScoreConfigString = $scoreConfigString;
+
 	async function refreshScore( force = false ) {
 		isLoading = true;
 		loadError = undefined;
 
 		try {
-			if ( force ) {
-				await clearCache();
-			}
-
-			scores = await requestSpeedScores();
-			scoreLetter = getScoreLetter( scores.mobile, scores.desktop );
+			scores = await requestSpeedScores( force );
+			scoreLetter = getScoreLetter( scores.current.mobile, scores.current.desktop );
+			showPrevScores = didScoresImprove( scores );
+			currentScoreConfigString = $scoreConfigString;
 		} catch ( err ) {
+			console.log( err );
 			loadError = err;
 		} finally {
 			isLoading = false;
 		}
 	}
+
+	const debouncedRefreshScore = debounce( force => {
+		// If the configuration has changed, the previous speed score is no longer relevant. But we don't want to
+		// trigger a refresh while critical css is still generating.
+		if (
+			! $criticalCssStatus.generating &&
+			$modulesInSync &&
+			$scoreConfigString !== currentScoreConfigString
+		) {
+			refreshScore( force );
+		}
+	}, 2000 );
+
+	$: {
+		// Mentioning the stores here for change detection.
+		$scoreConfigString;
+		$criticalCssStatus.generating;
+		$modulesInSync;
+
+		debouncedRefreshScore( true );
+	}
```

Comment:
This is pretty good, and I like your approach to explicitly mentioning variables inside the `$:` block. I do have a counter-idea - see what you think:

```suggestion
	const needRefresh = derived( [ criticalCssStatus, modulesInSync, scoreConfigString ] =>
		! $criticalCssStatus.generating &&
		$modulesInSync &&
		$scoreConfigString !== currentScoreConfigString
	);

	const debouncedRefreshScore = debounce( force => {
		if ( $needRefresh ) {
			refreshScore( force );
		}
	} );

	$: if ( $needRefresh ) {
		debouncedRefreshScore( true );
	}
```

I haven't tested this idea, but IMHO it keeps the relationships clear and simple. It ensures debouce is called whenever a need for refreshing arises, and double-checks that it is still needed at the last minute before actually doing the refresh. What do you think?

Reply:
I think your idea is neat. I have taken it in with some fixes.